File: 09-implementation/frontend/lib/core/models/submission.dart
class SubmissionVersionModel {
  final int submissionVersionId;
  final int submissionId;
  final int versionNo;
  final String? title;
  final String? externalLink;
  final String? textAnswer;
  final String? note;
  final DateTime submittedAt;

  SubmissionVersionModel({
    required this.submissionVersionId,
    required this.submissionId,
    required this.versionNo,
    this.title,
    this.externalLink,
    this.textAnswer,
    this.note,
    required this.submittedAt,
  });

  factory SubmissionVersionModel.fromJson(Map<String, dynamic> j) =>
      SubmissionVersionModel(
        submissionVersionId: j['submission_version_id'],
        submissionId: j['submission_id'],
        versionNo: j['version_no'],
        title: j['title'],
        externalLink: j['external_link'],
        textAnswer: j['text_answer'],
        note: j['note'],
        submittedAt: DateTime.parse(j['submitted_at']),
      );
}

class SubmissionDetail {
  final int submissionId;
  final int roundId;
  final int entryId;
  final int currentVersionNo;
  final String status;
  final bool isLocked;
  final DateTime? submittedAt;
  final List<SubmissionVersionModel> versions;

  SubmissionDetail({
    required this.submissionId,
    required this.roundId,
    required this.entryId,
    required this.currentVersionNo,
    required this.status,
    required this.isLocked,
    this.submittedAt,
    required this.versions,
  });

  factory SubmissionDetail.fromJson(Map<String, dynamic> j) => SubmissionDetail(
        submissionId: j['submission_id'],
        roundId: j['round_id'],
        entryId: j['entry_id'],
        currentVersionNo: j['current_version_no'],
        status: j['status'],
        isLocked: j['is_locked'],
        submittedAt: j['submitted_at'] != null ? DateTime.parse(j['submitted_at']) : null,
        versions: (j['versions'] as List? ?? [])
            .map((v) => SubmissionVersionModel.fromJson(v))
            .toList(),
      );
}

File: 09-implementation/backend/app/schemas/submission.py
"""Pydantic schemas cho submission (SV-08 nộp bài, GV-04 quản lý)."""

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.models.enums import SubmissionStatus


class SubmissionVersionCreateIn(BaseModel):
    """SV-08 POST /api/rounds/{id}/submissions/me/versions.

    SV nộp 1 version mới (text + link). Backend tự tạo submission nếu chưa.
    File upload chưa implement (TODO).
    """

    title: str | None = Field(None, max_length=255)
    description: str | None = None
    external_link: str | None = Field(None, description="URL Drive/GitHub")
    text_answer: str | None = None
    note: str | None = None


class SubmissionVersionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    submission_version_id: int
    submission_id: int
    version_no: int
    title: str | None = None
    description: str | None = None
    external_link: str | None = None
    text_answer: str | None = None
    checksum_value: str | None = None
    submitted_by: int | None = None
    submitted_at: datetime
    note: str | None = None


class SubmissionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    submission_id: int
    round_id: int
    entry_id: int
    current_version_no: int
    status: SubmissionStatus
    is_locked: bool
    submitted_at: datetime | None = None
    created_by: int | None = None
    updated_by: int | None = None
    created_at: datetime
    updated_at: datetime


class SubmissionDetail(SubmissionOut):
    versions: list[SubmissionVersionOut] = []
